File: PYTHON-IMPLEMENTATION/CODE/z.py
# Fills Z array for given string str[]
def getZarr(string, z):
    n = len(string)
    l, r, k = 0, 0, 0
    for i in range(1, n):
        if i > r:
            l, r = i, i
            while r < n and string[r - l] == string[r]:
                r += 1
            z[i] = r - l
            r -= 1
        else:
            k = i - l
            if z[k] < r - i + 1:
                z[i] = z[k]
            else:
                l = i
                while r < n and string[r - l] == string[r]:
                    r += 1
                z[i] = r - l
                r -= 1

# prints all occurrences of pattern 
# in text using Z algo
def search(text, pattern):
    concat = pattern + "$" + text
    l = len(concat)
    z = [0] * l
    getZarr(concat, z)
    for i in range(l):
        if z[i] == len(pattern):
            print("Pattern found at index", i - len(pattern) - 1)

if __name__ == "__main__":
    file_path = input("Enter the path to the file containing both text and pattern: ")

    try:
        with open(file_path, 'r') as file:
            content = file.read().strip().split('\n')
            text = content[0]
            pattern = content[1]

        search(text, pattern)

    except FileNotFoundError:
        print("File not found. Please check the file path.")

File: C++ IMPLEMENTATION/CODE/z.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>

using namespace std;

// Fills Z array for given string str[]
void getZarr(const string& str, vector<int>& z) {
    int n = str.length();
    int l = 0, r = 0, k = 0;
    for (int i = 1; i < n; ++i) {
        if (i > r) {
            l = r = i;
            while (r < n && str[r - l] == str[r])
                r++;
            z[i] = r - l;
            r--;
        } else {
            k = i - l;
            if (z[k] < r - i + 1)
                z[i] = z[k];
            else {
                l = i;
                while (r < n && str[r - l] == str[r])
                    r++;
                z[i] = r - l;
                r--;
            }
        }
    }
}

// Prints all occurrences of pattern in text using Z algorithm
void search(const string& text, const string& pattern) {
    string concat = pattern + "$" + text;
    int l = concat.length();
    vector<int> z(l, 0);
    getZarr(concat, z);
    for (int i = 0; i < l; ++i) {
        if (z[i] == pattern.length())
            cout << "Pattern found at index " << i - pattern.length() - 1 << endl;
    }
}

int main() {
    string file_path;
    cout << "Enter the path to the file containing both text and pattern: ";
    cin >> file_path;

    ifstream file(file_path);
    if (!file.is_open()) {
        cout << "File not found. Please check the file path." << endl;
        return 1;
    }

    string text, pattern;
    getline(file, text); // Assuming the first line contains text
    getline(file, pattern); // Assuming the second line contains pattern

    search(text, pattern);

    return 0;
}
